File: app/src/main/java/dev/eyosiyas/smsblocker/service/BlockerService.java
package dev.eyosiyas.smsblocker.service;

import android.content.BroadcastReceiver;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.telephony.SmsMessage;

import androidx.core.app.NotificationCompat;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

import dev.eyosiyas.smsblocker.database.DatabaseManager;
import dev.eyosiyas.smsblocker.model.Blacklist;
import dev.eyosiyas.smsblocker.model.Keyword;
import dev.eyosiyas.smsblocker.util.Core;
import dev.eyosiyas.smsblocker.util.InformManager;
import dev.eyosiyas.smsblocker.util.PrefManager;

import static dev.eyosiyas.smsblocker.util.Constant.CONTENT_PROVIDER_INBOX;
import static dev.eyosiyas.smsblocker.util.Constant.FIELD_BODY;
import static dev.eyosiyas.smsblocker.util.Constant.FIELD_DATE;
import static dev.eyosiyas.smsblocker.util.Constant.FIELD_NAME;
import static dev.eyosiyas.smsblocker.util.Constant.SMS_BUNDLE;

public class BlockerService extends BroadcastReceiver {
    private DatabaseManager databaseManager;

    @Override
    public void onReceive(Context context, Intent intent) {
        StringBuilder body = new StringBuilder();
        String number = "";
        long timestamp = System.currentTimeMillis();
        Bundle bundle = intent.getExtras();
        SmsMessage[] messages;
        if (bundle != null) {
            databaseManager = new DatabaseManager(context);
            Object[] msgObjects = (Object[]) bundle.get(SMS_BUNDLE);
            messages = new SmsMessage[msgObjects.length];
            for (int i = 0; i < messages.length; i++) {
                messages[i] = SmsMessage.createFromPdu((byte[]) msgObjects[i]);
                body.append(messages[i].getMessageBody());
                number = messages[i].getOriginatingAddress();
                timestamp = messages[i].getTimestampMillis();
            }
            createMessage(context, number, body.toString(), timestamp);
        }
    }

    private void createMessage(Context context, String address, String body, long timestamp) {
        PrefManager manager = new PrefManager(context);
        boolean isClean = true;
        for (Blacklist blacklist : databaseManager.getBlacklist()) {
            if (address.equalsIgnoreCase(blacklist.getNumber())) {
                isClean = false;
                break;
            } else
                isClean = true;
        }

        if (isClean && manager.isNuclearEnabled()) {
            Pattern nuclearPattern = Pattern.compile("(?<!\\d)\\d{3,4}(?!\\d)");
            Matcher matcher = nuclearPattern.matcher(address);
            isClean = !matcher.find();
        }

        if (isClean && manager.isStartsWithEnabled()) {
            Pattern startsWithPattern = Pattern.compile(String.format("^%s", manager.getStartsWith()));
            Matcher matcher = startsWithPattern.matcher(address);
            isClean = !matcher.find();
        }

        if (isClean && manager.isEndsWithEnabled()) {
            Pattern endsWithPattern = Pattern.compile(String.format("%s$", manager.getEndsWith()));
            Matcher matcher = endsWithPattern.matcher(address);
            isClean = !matcher.find();
        }

        if (isClean && databaseManager.getKeywordsCount() > 0) {
            for (Keyword keyword : databaseManager.getKeywords()) {
                if (body.contains(keyword.getKeyword())) {
                    isClean = false;
                    break;
                } else {
                    isClean = true;
                }
            }
        }

        if (isClean) {
            ContentValues contentValues = new ContentValues();
            contentValues.put(FIELD_NAME, address);
            contentValues.put(FIELD_BODY, body);
            contentValues.put(FIELD_DATE, timestamp);
            ContentResolver contentResolver = context.getContentResolver();
            contentResolver.insert(CONTENT_PROVIDER_INBOX, contentValues);
            notifyUser(context, address, body);
        }
    }

    private void notifyUser(Context context, String number, String message) {
        InformManager inform = new InformManager(context);
        NotificationCompat.Builder builder = inform.showNotification(Core.getDisplayName(context, number), message);
        inform.getManager().notify((int) (1 + Math.random() * 9999), builder.build());
    }
}

File: app/src/main/java/dev/eyosiyas/smsblocker/service/BlockerService.kt
package dev.eyosiyas.smsblocker.service

import android.content.*
import android.os.Build
import android.os.Bundle
import android.telephony.SmsMessage
import androidx.core.app.NotificationCompat
import dev.eyosiyas.smsblocker.database.Storage
import dev.eyosiyas.smsblocker.model.Blacklist
import dev.eyosiyas.smsblocker.model.Blocked
import dev.eyosiyas.smsblocker.model.Keyword
import dev.eyosiyas.smsblocker.model.Whitelist
import dev.eyosiyas.smsblocker.util.Constant
import dev.eyosiyas.smsblocker.util.Constant.SMS_FORMAT
import dev.eyosiyas.smsblocker.util.Core
import dev.eyosiyas.smsblocker.util.InformManager
import dev.eyosiyas.smsblocker.util.PrefManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.util.regex.Matcher
import java.util.regex.Pattern


class BlockerService : BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        val body: StringBuilder = StringBuilder()
        var number = ""
        var timestamp: Long = System.currentTimeMillis()
        val bundle: Bundle? = intent.extras
        val messages: Array<SmsMessage?>

        val format = bundle!!.getString(SMS_FORMAT)
        val msgObjects: Array<*>? = bundle.get(Constant.SMS_BUNDLE) as Array<*>?
        messages = arrayOfNulls(msgObjects!!.size)
        for (i in messages.indices) {
            messages[i] = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
                SmsMessage.createFromPdu(msgObjects[i] as ByteArray, format)
            else
                SmsMessage.createFromPdu(msgObjects[i] as ByteArray)
            body.append(messages[i]!!.messageBody)
            number = messages[i]!!.originatingAddress.toString()
            timestamp = messages[i]!!.timestampMillis
        }
//            for (i in messages.indices) {
//                messages[i] = SmsMessage.createFromPdu(msgObjects[i] as ByteArray?)
//                body.append(messages[i]!!.messageBody)
//                number = messages[i]!!.originatingAddress.toString()
//                timestamp = messages[i]!!.timestampMillis
//            }
        GlobalScope.launch(Dispatchers.IO) {
            createMessage(context, number, body.toString(), timestamp)
        }
    }

    private suspend fun createMessage(context: Context, address: String, body: String, timestamp: Long) {
        val db = Storage.database(context)
        val blacklistDAO = db.blacklistDao()
        val blockedDAO = db.blockedDao()
        val keywordDAO = db.keywordDao()
        val whitelistDAO = db.whitelistDao()
        val manager = PrefManager(context)
        var isClean = true

        if (blacklistDAO.blacklistCount() > 0) {
            for (blacklist: Blacklist in blacklistDAO.pureBlacklists()) {
                if (address.equals(blacklist.number, ignoreCase = true)) {
                    isClean = false
                    break
                }
            }
        }

        if (isClean && keywordDAO.keywordsCount() > 0) {
            for (keyword: Keyword in keywordDAO.pureKeywords()) {
                if (body.contains(keyword.keyword)) {
                    isClean = false
                    break
                }
            }
        }

        if (isClean) {
            var matcher: Matcher
            when (manager.blockingRule) {
                Constant.STARTS_WITH_TAG -> {
                    val startsWithPattern: Pattern = Pattern.compile(String.format("^%s", manager.startsWith))
                    if (whitelistDAO.whitelistCount() > 0) {
                        for (whitelist: Whitelist in whitelistDAO.pureWhitelists()) {
                            if (!address.equals(whitelist.number, ignoreCase = true)) {
                                matcher = startsWithPattern.matcher(address)
                                if (matcher.find()) {
                                    isClean = false
                                    break
                                }
                            }
                        }
                    } else
                        isClean = !startsWithPattern.matcher(address).find()
                }
                Constant.ENDS_WITH_TAG -> {
                    val endsWithPattern: Pattern = Pattern.compile(String.format("%s$", manager.endsWith))
                    if (whitelistDAO.whitelistCount() > 0) {
                        for (whitelist: Whitelist in whitelistDAO.pureWhitelists()) {
                            if (!address.equals(whitelist.number, ignoreCase = true)) {
                                matcher = endsWithPattern.matcher(address)
                                if (matcher.find()) {
                                    isClean = false
                                    break
                                }
                            }
                        }
                    } else
                        isClean = !endsWithPattern.matcher(address).find()
                }
                Constant.NUCLEAR_OPTION_TAG -> {
                    val nuclearPattern: Pattern = Pattern.compile("(?<!\\d)\\d{3,4}(?!\\d)")
                    if (whitelistDAO.whitelistCount() > 0) {
                        for (whitelist: Whitelist in whitelistDAO.pureWhitelists()) {
                            if (!address.equals(whitelist.number, ignoreCase = true)) {
                                matcher = nuclearPattern.matcher(address)
                                if (matcher.find()) {
                                    isClean = false
                                    break
                                }
                            }
                        }
                    } else
                        isClean = !nuclearPattern.matcher(address).find()
                }
            }
        }

        if (isClean) {
            val contentValues = ContentValues()
            contentValues.put(Constant.FIELD_NAME, address)
            contentValues.put(Constant.FIELD_BODY, body)
            contentValues.put(Constant.FIELD_DATE, timestamp)
            val contentResolver: ContentResolver = context.contentResolver
            contentResolver.insert(Constant.CONTENT_PROVIDER_INBOX, contentValues)
            notifyUser(context, address, body)
        } else
            blockedDAO.addBlocked(Blocked(0, address, body, timestamp))
    }

    private fun notifyUser(context: Context, number: String, message: String) {
        val inform = InformManager(context)
        val builder: NotificationCompat.Builder = inform.showNotification(Core.displayName(context, number), message)
        inform.manager.notify((1..9999).random(), builder.build())
    }
}